### PR TITLE
Private instances

### PR DESCRIPTION
**Pull request by thebaer**
This changes the `private` config setting to make an instance entirely private. That is, when enabled, blogs will only be readable by other authenticated users on the same instance.

It still needs a bit of testing, so it's a draft for now.

When finished, this closes [T576](https://phabricator.write.as/T576).

### CODE REVIEW

File: posts.go [hunk 1]
```diff
@@ -1193,6 +1193,16 @@ func getRawCollectionPost(app *App, slug, collAlias string) *RawPost {
...
+	slug := vars["slug"]
+
+	isJSON := strings.HasSuffix(slug, ".json")
+	isXML := strings.HasSuffix(slug, ".xml")
+	isMarkdown := strings.HasSuffix(slug, ".md")
+	return strings.HasSuffix(slug, ".txt") || isJSON || isXML || isMarkdown
+}
+
 func viewCollectionPost(app *App, w http.ResponseWriter, r *http.Request) error {
 	vars := mux.Vars(r)
```
Comment (robjloranger) [inline posts.go]: these 7 lines are redundant now with `isRaw` above.

File: posts.go [hunk 2]
```diff
@@ -1202,12 +1212,6 @@ func viewCollectionPost(app *App, w http.ResponseWriter, r *http.Request) error
 	isMarkdown := strings.HasSuffix(slug, ".md")
 	isRaw := strings.HasSuffix(slug, ".txt") || isJSON || isXML || isMarkdown
 
-	if strings.Contains(r.URL.Path, ".") && !isRaw {
```
Comment (robjloranger) [inline posts.go]: should this still check if `isRaw`?

File: handle.go [hunk 3]
```diff
@@ -331,21 +386,21 @@ func (h *Handler) WebErrors(f handlerFunc, ul UserLevel) http.HandlerFunc {
 
 			var session *sessions.Session
 			var err error
-			if ul != UserLevelNone {
+			if ul(h.app.App().cfg) != UserLevelNoneType {
 				session, err = h.sessionStore.Get(r, cookieName)
-				if err != nil && (ul == UserLevelNoneRequired || ul == UserLevelUser) {
+				if err != nil && (ul(h.app.App().cfg) == UserLevelNoneRequiredType || ul(h.app.App().cfg) == UserLevelUserType) {
 					// Cookie is required, but we can ignore this error
 					log.Error("Handler: Unable to get session (for user permission %d); ignoring: %v", ul, err)
```
Comment (robjloranger) [inline handle.go]: need to call `ul()` here

File: handle.go [hunk 4]
```diff
@@ -404,21 +475,21 @@ func (h *Handler) Web(f handlerFunc, ul UserLevel) http.HandlerFunc {
 				log.Info("\"%s %s\" %d %s \"%s\"", r.Method, r.RequestURI, status, time.Since(start), r.UserAgent())
 			}()
 
-			if ul != UserLevelNone {
+			if ul(h.app.App().cfg) != UserLevelNoneType {
 				session, err := h.sessionStore.Get(r, cookieName)
-				if err != nil && (ul == UserLevelNoneRequired || ul == UserLevelUser) {
+				if err != nil && (ul(h.app.App().cfg) == UserLevelNoneRequiredType || ul(h.app.App().cfg) == UserLevelUserType) {
 					// Cookie is required, but we can ignore this error
 					log.Error("Handler: Unable to get session (for user permission %d); ignoring: %v", ul, err)
```
Comment (robjloranger) [inline handle.go]: again here, call `ul()`

File: handle.go [hunk 5]
```diff
@@ -523,27 +652,27 @@ func (h *Handler) Download(f dataHandlerFunc, ul UserLevel) http.HandlerFunc {
...
 			start := time.Now()
 
 			var status int
-			if ul != UserLevelNone {
+			if ul(h.app.App().cfg) != UserLevelNoneType {
 				session, err := h.sessionStore.Get(r, cookieName)
-				if err != nil && (ul == UserLevelNoneRequired || ul == UserLevelUser) {
+				if err != nil && (ul(h.app.App().cfg) == UserLevelNoneRequiredType || ul(h.app.App().cfg) == UserLevelUserType) {
 					// Cookie is required, but we can ignore this error
 					log.Error("Handler: Unable to get session (for user permission %d); ignoring: %v", ul, err)
```
Comment (robjloranger) [inline handle.go]: again here, `ul()`

File: handle.go [hunk 6]
```diff
@@ -23,24 +23,52 @@ import (
...
-	UserLevelOptional                      // user or not -- object fetched if user
-	UserLevelNoneRequired                  // non-user (required)
-	UserLevelUser                          // user (required)
+	UserLevelNoneType         UserLevel = iota // user or not -- ignored
+	UserLevelOptionalType                      // user or not -- object fetched if user
+	UserLevelNoneRequiredType                  // non-user (required)
+	UserLevelUserType                          // user (required)
 )
 
+func UserLevelNone(cfg *config.Config) UserLevel {
```
[the remark this answers is not in the document]
Comment (thebaer) [reply handle.go]: Yeah, to satisfy the `UserLevelFunc` type from line 71. The `UserLevelReader()` func does use the config

File: posts.go (hunk 2 above)
Comment (thebaer) [reply posts.go]: The condition will always be false here, as it's now handled in `CollectionPostOrStatic()` in `handle.go`. The move to an additional func was needed to always allow access to static assets, while restricting access to posts if necessary.

File: posts.go (hunk 1 above)
Comment (thebaer) [reply posts.go]: I know it's not the cleanest, but we need the result of `isRaw()` for `CollectionPostOrStatic()`, and we need all the vars in that logic below. I put them close together so we could have both without a ton of refactoring. Maybe for now I should add some comments mentioning that the two places need to be kept in parity?

File: handle.go (hunk 3 above)
Comment (thebaer) [reply handle.go]: Good catch -- will fix these.

Comment (thebaer) [reply handle.go]: Done.

File: posts.go (hunk 1 above)
Comment (robjloranger) [reply posts.go]: sure that's probably good

**Comment by robjloranger** [comment]
testing looks good running locally. I haven't tested federation though, did you want me to do that too?

File: posts.go (hunk 1 above)
Comment (thebaer) [reply posts.go]: done :+1:

**Comment by thebaer** [comment]
Thanks @robjloranger -- no need to test federation. Merging now!

**Pull request merged**